src/peergos/server/storage/S3Request.java

```diff
@@ -0,0 +1,204 @@
+                + ",SignedHeaders=" + headersToSign() + ",Signature=" + signature);
+        return headers;
+    }
+
+    public Map<String, String> getOriginalHeaders() {
+        Map<String, String> res = new LinkedHashMap<>();
+        res.put("Host", host);
+        res.put("x-amz-date", asAwsDate(date));
+        res.put("x-amz-content-sha256", contentSha256);
+        for (Map.Entry<String, String> e : extraHeaders.entrySet()) {
+            res.put(e.getKey(), e.getValue());
+        }
+        if (allowPublicReads)
+            res.put("x-amz-acl", "public-read");
+        return res;
+    }
+
+    public SortedMap<String, String> getCanonicalHeaders() {
+        SortedMap<String, String> res = new TreeMap<>();
+        Map<String, String> originalHeaders = getOriginalHeaders();
+        for (Map.Entry<String, String> e : originalHeaders.entrySet()) {
+            res.put(e.getKey().toLowerCase(), e.getValue());
+        }
+        return res;
+    }
+
+    public String headersToSign() {
+        return getCanonicalHeaders().keySet()
+                .stream()
+                .sorted()
+                .collect(Collectors.joining(";"));
+    }
+
+    public String toCanonicalRequest() {
+        StringBuilder res = new StringBuilder();
+        res.append(verb + "\n");
+        res.append("/" + key + "\n");
+
+        res.append("\n"); // no query parameters
+
+        Map<String, String> headers = getCanonicalHeaders();
+        for (Map.Entry<String, String> e : headers.entrySet()) {
+            res.append(e.getKey().toLowerCase() + ":" + e.getValue() + "\n");
+        }
+        res.append("\n");
+
+        res.append(headersToSign() + "\n");
+        res.append(contentSha256);
+        return res.toString();
+    }
+
+    private String scope() {
+        return String.format(
+                "%s/%s/%s/%s",
+                asAwsShortDate(date),
+                region,
+                "s3",
+                "aws4_request");
+    }
+
+    public String stringToSign() {
+        StringBuilder res = new StringBuilder();
+        res.append("AWS4-HMAC-SHA256" + "\n");
+        res.append(asAwsDate(date) + "\n");
+        res.append(scope() + "\n");
+        res.append(ArrayOps.bytesToHex(sha256(toCanonicalRequest().getBytes())));
+        return res.toString();
+    }
+
+    public static String asAwsDate(Instant instant) {
+        return instant.toString()
+                .replaceAll("[:\\-]|\\.\\d{3}", "");
+    }
+
+    public static String asAwsShortDate(Instant instant) {
+        return asAwsDate(instant).substring(0, 8);
+    }
+
+    public final String verb, host;
+    public final String key;
+    public final String contentSha256;
+    public final boolean allowPublicReads;
+    public final String accessKeyId;
+    public final String region;
+    public final Map<String, String> extraHeaders;
+    public final Instant date;
+
+    public S3Request(String verb,
```

Comment:
Feels weird the member vars and constructor at the end of file

Reply:
Agreed, that was an accident. Fixed.